만국박람회 [STEP 2] Taeangel, dudu

FirstDo opened the pull request:
안녕하세요 린생 @jungseungyeo , 만국박람회 STEP2 PR 보냅니다 
잘부탁드립니다 :)

### 고민한점

#### name에 string literal이 그대로 들어가는 문제

`"items"`이나 `"exposition_universelle_1900"`에 string literal를 그대로 사용해도 되는지 아니면 관리를 하여 사용해서 사용하는 것중 어느것이 좋은지 궁금합니다.

#### UIView의 SubViews 요소의 선언 위치

예를들어서 HeritageDetailViewController에는

heritageImageView와 descriptionLabel 요 두개가 필요한데요,
굳이 프로퍼티로 선언을 하지 않아도, layout 이라는 함수 내부에서 지역변수로 생성해서 데이터를 넣어줘도 되더라구요

먼가, 더 유연한 설계를 위해서 프로퍼티로 빼놓긴 했는데, 이경우는 그냥 지역변수로 생성해서 사용해도 큰 문제가 없는걸까요?

#### `개최지: 프랑스파리` 를 표현하는데 레이블을 2개로 할지, `attributedText`를 사용해볼지

`개최지 : 프랑스 파리` 를 출력할때, 개최지만 다른 폰트를 가지고 있어서 해당 문제를 어떻게 해결할지 고민했습니다.

저희가 생각하기로는

1) 개최지 레이블 따로, 프랑스 파리 레이블 따로
2) 하나의 레이블로 attributedText 이용

의 방법이 있었고, 2번방법을 한번도 사용해 본적 없어서, 같이 공부후, 적용해보았습니다

근데 굳이 따지면, 레이블을 두개 넣는 방법이 더 좋은거 같기도한데.. 여기에 대한 생각이 궁금합니다


#### tableView에서 list가 필요해서 선언할때

예를들어 HeritageViewController에는 TableView에 표시할 데이터를 담아둘 배열인 heritageList가 있는데,
해당 리스트에는 json 파일로부터 decode한 [Heritage]가 들어가야 합니다

저희가 고민한게

1. 아예 프로퍼티 기본값을 주는 방법
    private let list = ParserManager<[Heritage]>.parser
2. 옵셔널로 선언한 다음, 추후 초기화 해주는 방법
    private var list: [Heritage]?
4. 빈 배열로 선언한 다음, 추후 초히과 해주는 방법
    private var list = [Heritage]()

저희는 3번이 제일 좋은것 같아서 선택했는데
(굳이 옵셔널로 선언할 필요 없음, 프로퍼티 기본값을 주는게 조금 불편함)
어떤방식이 좋을지 결론을 내리지 못했습니다

#### required init fatal error

코드 베이스라서 어차피 `required init`이 불릴일은 절대 없다고 생각해서 fatalError()를 사용했는데 이래도 되는걸까요? 아니면 더 좋은 방법이 있을까요??

#### TableViewCell의 reuseIdentifier

문자열을 그대로 쓰는걸 피하기 위해, 예전 쥬스메이커에서 ViewController의 Storyboard Identifer를 
아래와 같이 설정했었습니다

```swift
extension UIViewController {

  static var identifier: String {
    return String(describing: self)
  }
}
```

TableViewCell의 reuseIdentifier도 해당 방식과 비슷하게 해보았는데, 괜찮은 방법일까요? 혹시 더 좋은 방법이나 린생이 사용하시는 방법이 궁금합니다

```swift
extension UITableViewCell {

  static var identifier: String {
    return String(describing: self)
  }
}
```

#### 프로젝트 파일 그룹화에 관해서

파일을 관리할때 App Delegate, Scene Delegate도 따로 폴더에 넣어주는게 좋을까요..? 넣는다면 어떤이름이 좋을지 궁금합니다.. View도 Model도 Controller도 아닌거 같기도하고.. 또 Controller같기도 하고요

#### CellForRowAt 매서드의 return 

tableView에서 cell을 guard let을 이용해서 얻어오는데 실패했을때도 UITableViewCell을 반환해야 하는데, 현재는 UITableViewCell()을 그냥 반환해주고 있습니다. 해당 방식에 문제가 있거나 혹시 더 좋은 방법이 있을까요?

이 부분에 있어서 다른분들께도 질문을 드리고 생각해봤는데, UITableViewCell()을 바로 반환하는 방식이 별로 좋은것 같지는 않은거같아요.
결정적으로 UITableViewCell()을 그냥 반환해버리면, 모두 비어있는 Cell 몇개가 생성되는데, 해당 Cell은 여전히 didSelectRowAt 매서드가 있기때문에, 여전히 select 이벤트를 처리해서 다음화면으로 넘어가더라구요..!

잘못된 Cell일때는 selection을 막는 식으로 구현해보려고 하는데, 이런식으로 접근하는게 맞을까요?

Taeangel committed:
feat: 데이터를 얻기위한 함수 구현

Expo와 poster 이미지를 튜플로 반환하는 함수 구현

FirstDo committed:
Merge branch 'STEP1' into STEP2

FirstDo committed:
feat: scrollView와 stackView의 제약 생성

Taeangel committed:
feat: stackView 내부 구성요소 추가 및 layout 설정

FirstDo committed:
refactor: view 요소들을 프로퍼티로 변경, 레이아웃을 extension으로 분리

Taeangel committed:
feat: 프로퍼티에 데이터 설정

FirstDo committed:
fix: crash 나는 부분 수정 및 layout 수정

buttonStackView.addArrangedSubview(buttonStackView)를 삭제
StackView와 ScrollView의 여백 조정

Taeangel committed:
style: 컨벤션에 맞게 개행수정, 네이밍 변경

button을 koreaHeritageButton로 변경
imageView를 posterImageView로 변경

FirstDo committed:
feat: 요구조건에 맞게, Label을 수정하는 함수 구현

특정 문자열을 추가하고 NSAttributedString을 이용해서 해당 문자열의 크기 변경

Taeangel committed:
feat: addComma 메서드 구현

숫자사이에 ","와 숫자 끝에 "명"을 추가하는 메서드를 Int extension에 구현
changeTextColor 메서드 네이밍을 changeFontSize로 변경
changeFontSize 메서드를 String extension으로 이동

FirstDo committed:
feat: ExpoViewController를 네비게이션 컨트롤러에 embed

navigationBar를 hidden으로 설정
navigationBar의 backButtonTitle을 "메인" 으로 설정

Taeangel committed:
feat: herotageViewController로 이동하는 기능 구현

뷰라이프 사이클을 통해 navigationBar를 가리고 보여주는 기능 구현
HeritageViewController 생성 및 초기화

FirstDo committed:
feat: TableView를 생성하고 layout 설정

Taeangel committed:
feat: prepareData 메서드와 heritageList 구현

FirstDo committed:
feat: HeritageCell 구현

내부 view 생성 및 레이아웃 설정

Taeangel committed:
feat: update 메서드 구현, identifier 메서드 구현

FirstDo committed:
feat: TableView에 Cell을 등록하고 datasource 매서드 구현

Taeangel committed:
fix: HeritageCell의 제약조건 수정

FirstDo committed:
feat: HeritageDetailViewController 생성

custom initalizer 구현

Taeangel committed:
feat: tableViewDelegate 메서드 구현

cell을 눌렀을때  heritageDetailViewController를  navigationController에 푸쉬

FirstDo committed:
feat: HeritageDEtailViewController를 구성

subViews를 layout 설정 및 데이터 설정

Taeangel committed:
feat: Cell accessoryType에 indicator 추가

//MARK: - Layout 추가와 전체적인 return 인덴트 수정

FirstDo committed:
chore: layout constant의 매직 넘버 삭제

jungseungyeo commented on Expo1900/Expo1900/Controller/ExpoViewController.swift:
```diff
@@ -0,0 +1,220 @@
+    let numberFormatter = NumberFormatter()
+    numberFormatter.numberStyle = .decimal
+    guard let formattedNumber = numberFormatter.string(for: self) else {
+      return "0 명"
```
addComma라는 네이밍에서 return될거라고 예상 되는 값은 아니거 같네요.

Taeangel replied:
addComma 메서드를 `decimalStyleFormat`로 네이밍을 변경하고 `add(text: String)`으로 기능을 분리했습니다.

jungseungyeo commented on Expo1900/Expo1900/Controller/ExpoViewController.swift:
```diff
@@ -0,0 +1,220 @@
+    navigationItem.backButtonTitle = "메인"
+    view.backgroundColor = .systemBackground
+    
+    guard let (expo, poster) = prepareData() else {
```
모델을 따로 만들어주세요. 여기서 넣고 바로 하는게 아니라 모델을 주입하는 부분과 ui를 업데이트하는 부분을 분리하면 좋을거같아요.

jungseungyeo commented on Expo1900/Expo1900/Controller/ExpoViewController.swift:
```diff
@@ -0,0 +1,220 @@
+
+extension ExpoViewController {
+  
+  private func layout() {
```
layout함수에서 ui를 그려주는 것이 좋네요! 하지만 코드를 보았을때 정리 된 느낌이 없네요. 
뷰 하나하나 그리는 것으로 변경하면 좋을거같아요.

FirstDo replied:
ViewController와 View를 분리했습니다
View에서는 View를 그리는것만 담당하고,
Controller는 View와 Model을 Bind 해주는걸 담당하도록 바꿔보았습니다..!

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageDetailViewController.swift:
```diff
@@ -0,0 +1,88 @@
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    attribute()
+    layout()
```
모델을 주입후 layout을 그리면 좋지면 기본적으로 모델은 서버에서 받아오는 경우가 많죠? 
attribute와 layout의 순서를 변경하여 view를 그리는 방식을 고려해주시면 좋을거같아요.

FirstDo replied:
json Model을 파싱할때 서버에서 값을 받아오는 느낌으로 Result Type과 Completion Hanlder를 사용했습니다
models을 저장할 프로퍼티에 프로퍼티 옵져버를 활용하여, 모델이 업데이트 된 후, 명시적으로 view를 업데이트 하도록 변경하였습니다

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageViewController.swift:
```diff
@@ -0,0 +1,76 @@
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    attribute()
+    layout()
```
위의 리뷰에서 한거처럼 attribute와 layout의 순서를 변경하여 ui를 작성해주시면 좋을거같아요.

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    let heritage = heritageList[indexPath.row]
+    cell.update(with: heritage)
```
`cell.update(with: heritageList[indexPath.row])`와 위의 코드의 차이점이 무엇일까요?

FirstDo replied:
변수에 저장한후, 사용하려고 했는데
이 경우,  굳이 나눠줄 필요가 없는것 같네요..!
하나로 합치고, 링크주신 subscript의 extension을 활용해서 safe하게 배열을 조회하도록 변경하였습니다

jungseungyeo reviewed (changes requested):
### name에 string literal이 그대로 들어가는 문제

`따로 만들어서 관리한다는게 의미를 잘 이해가 안되는거 같아요. 따로 Model의 이야기라면 만들어서 관리하면 좋을거같습니다 :) `

### UIView의 SubViews 요소의 선언 위치

`지역변수로 해도 문제는 없지만  layout 함수가 커지는 문제가 있는거 같아요. 지역변수와 전역변수는 큰 문제가 되지는 않는거 같아요 ㅎㅎ`

### 개최지: 프랑스파리 를 표현하는데 레이블을 2개로 할지, attributedText를 사용해볼지, 개최지 : 프랑스 파리 를 출력할때, 개최지만 다른 폰트를 가지고 있어서 해당 문제를 어떻게 해결할지 고민했습니다.

`extension을 활용하여 적용한 부분 좋아요. 하지만 네이밍은 문제가 있는거 같아요 무엇을 change하는지 보다 명확히 표현을 해주면 좋을거같아요.  또 두 레이블을 합하는 방법과 하나를 적용하는 방법중 저는 관리가 편한 방법이 좋은거 같아요. attributedText 를 사용하면 적용하신거처럼 확장을 통해 기능을 추가해주는 방법이고 레이블 두개를 사용한다면 크게 문제는 없겟죠?`

### tableView에서 list가 필요해서 선언할때

`이 방법은 2번을 사용하는 방법을 추천드려요. 현재는 데이터를 로컬에서 가져오는 경우지만 대부분 서버에서 값을 받아오는거죠? 서버에서 받아오는 경우 model을 못 가져오는 경우도 있기 때문에 안전하게 2번의 방법을 추천합니다.`

### required init fatal error

`저도 많이 사용합니다. 코드베이스인경우 코드가 아닌 방법으로 진입시 확실하게 에러를 하는 방법이 저는 좋지만 좋게 보지 않는 사람도 있어요. 이부분은 유연하게 받아드리면 좋을거같아요. 하지만 저는 이렇게 사용하는 방법을 좋아합니다.`

### TableViewCell의 reuseIdentifier

`이런 형태로 저도 많이 사용해요! 너무 좋습니다.`

### 프로젝트 파일 그룹화에 관해서

`저는 Entry라는 폴더 안에 넣어두는 편입니다. Scene이라는 폴더안에도 넣어두는 편이예요. 자신의 기준을 잡고 어느 폴더에서 관리하면 좋을지 고민해보세요 :) `

### CellForRowAt 매서드의 return

`흠.. 대부분 guard let 으로 가져오는 편입니다. 하지만 모델배열에서 안전하게 가져오는 방법이 있습니다.`
http://minsone.github.io/programming/check-index-of-array-in-swift
위 링크를 참조해주세요 :) 

`또 cell UITableViewCell이 호출되면 앱이 크래시가 발생할겁니다 :) `

태엔젤, 두두 디코에서 이야기한것처럼 수정을 중점으로해주시면 좋을거같아요. 고생하셨습니다.

Taeangel committed:
refactor: ExpoViewController를 View와 Controller로 분리

FirstDo committed:
refactor: model을 업데이트 한 후, view가 업데이트 되도록 변경

Taeangel committed:
refactor:  addComma 메서드 변경

decimalStyleFormat 메서드와 personFormat 메서드로 분리

FirstDo committed:
refactor: "방문객 : ", "개최지 : ", "개최 기간 : " 표현 방식 변경

attributeText를 이용하던 방식에서 Label하나 더 추가해서 표현하는 방식으로 변경

Taeangel committed:
feat: addSubView와 addArrangedSubView를 기능확장

addSubView와 addArrangedSubView 기능을 가변 전달인자를 활용하여 확장

FirstDo committed:
refactor: addArrangedSubviews 매서드 사용

Taeangel committed:
style: Const와 layout을 extention으로 분리

atrribute 메서드생성
전체적인 개행 변경
매직넘버 삭제

FirstDo committed:
refactor: String extension에 매서드를 만들어서 기능분리

타겟을 기준으로 라인을 분리하는 divideLine(text: ) 매서드 생성
personFormat을 add(text: String) 으로 변경

Taeangel committed:
style: 매직넘버 삭제

FirstDo committed:
feat: HeritageViewController를 View와 Controller로 분리

Taeangel committed:
style: 매직넘버 삭제

Taeangel committed:
Merge remote-tracking branch 'refs/remotes/origin/STEP2'

FirstDo committed:
style: HeritageCell의 매직 넘버 삭제

FirstDo committed:
Merge branch 'STEP2' of https://github.com/Taeangel/ios-exposition-universelle into STEP2

Taeangel committed:
refactor: layout 기능 분리

FirstDo committed:
fix: addSubviews 오타로 인한 버그 수정

Taeangel committed:
refactor: HeritageViewController를 View와 Controller로 분리

FirstDo committed:
style: HeritageDetailViewController, View의 매직넘버 삭제

마크다운 주석 추가

Taeangel committed:
fix: HeritageViewController의 잘못된 title 수정

FirstDo committed:
refactor: 기존 parse를 Decodable의 extension으로 변경

error 처리 추가
기존 테스트 코드 수정

Taeangel committed:
style: View와 Controller를 그룹화

FirstDo committed:
style: final 키워드 추가, 프로퍼티에 private lazy var 키워드 추가

Taeangel committed:
feat: extension Array를 하여 safe subscript 구현

FirstDo committed:
fix: ScrollBar 위치 변경

FirstDo commented:
말씀해주신부분 위주로 고쳐보았는데, 잘 된건지 모르겠네요..!!

- view와 controller를 분리
- MVC 구조 준수를 위해서, Model을 업데이트 될때, View가 업데이트 되도록 구성
- Array의 safe subscript 구현
- class에 final 키워드 추가
- Const 네임스페이스로 매직 넘버/리터럴 관리
- MARK: -, extension을 이용해서 가독성 높임
- ParserManager 삭제 후 Decodable Extension으로 확장 (ResultType과 Completion Handler를 사용해서 비동기 느낌처럼 수정)
- 기타 버그수정

#### 궁금한점: 

현재는 ViewController에 bind() 매서드가 있어서, VC가 View에 데이터를 넣어줍니다.
근데 이런방식은 은닉화 / 캡슐화를 위반하는것이지만
bind() 매서드가 View에 있는건 좀 이상하다고 생각하고, View는 진짜 보여지는것만 (layout) 담당해야 한다고 생각해서 현재처럼 구현했습니다.
저번에 두 방식 모두 괜찮다고 설명해 주신것 같아서요 :)

부족한 부분 너무 많으니 잘못된 부분이나 개선해야할 부분이 있다면 바로 피드백 부탁드립니다 :)

Taeangel committed:
docs:

jungseungyeo commented on Expo1900/Expo1900/Extention/Decodable+.swift:
```diff
@@ -0,0 +1,29 @@
+      completion(.success(decodedData))
+    } catch {
+      completion(.failure(MyError.decodeFail))
+    }
```
여기서 처리 할 필요는 없는거 같아요.. 이러면 model이 작업하는 느낌이네요.. controller로 옮기면 좋을거같아요.

jungseungyeo commented on Expo1900/Expo1900/Extention/UIStackView+.swift:
```diff
@@ -0,0 +1,16 @@
+import UIKit
+
+extension UIStackView {
+  func addArrangedSubviews(_ views: UIView...) {
```
좋습니다 :)

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageViewController/HeritageViewController.swift:
```diff
@@ -0,0 +1,98 @@
+  }
+  
+  private func requestData() {
+    [Heritage].parse(name: Const.File.name) { result in
```
흠.. parse함수가 아닌 requestDatark complete가 존재해야 될거같네요. model에서 실패 성공 여부를 판단하는 모양이 되어 버린거 같네요 ㅎㅎ.. 다음 step에서 수정해주세요

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageViewController/HeritageCell.swift:
```diff
@@ -0,0 +1,121 @@
+    
+    //MARK: - heritageImageView
+    
+    NSLayoutConstraint.activate([
```
완전 제 스타일이네요 ㅎㅎ... 너무 똑같아진거 같지만 보기 좋네요 ㅎㅎ.. 앞으로 스스로의 스타일을 가지면 더욱 좋을거같아요 ㅎㅎ 전 보다 확실히 코드 베이스의 view가 그려지는 모양이 잘 보여지는거 같아요

jungseungyeo commented on Expo1900/Expo1900/Controller/ExpoViewController/ExpoViewController.swift:
```diff
@@ -0,0 +1,121 @@
+  }
+  
+  private func requestData() {
+    Expo.parse(name: Const.File.name) { result in
```
여기도 parse보다는 requestData가 complete를 가지고 있어야 될거같네요 ㅠ

jungseungyeo commented on Expo1900/Expo1900/Controller/ExpoViewController/ExpoViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+  }
+  
+  private func bind() {
```
model을 넣어주는 부분과 bind를 해주는 부분이 분리가 되어서 좋네요 ㅎㅎ

jungseungyeo commented on Expo1900/Expo1900/Extention/Array+.swift:
```diff
@@ -0,0 +1,12 @@
+//
+
+extension Array {
+  subscript(safe index: Int) -> Element? {
```
너무 좋습니다.

jungseungyeo commented on Expo1900/Expo1900/Controller/HeritageViewController/HeritageViewController.swift:
```diff
@@ -0,0 +1,98 @@
+  private var heritageList: [Heritage]? {
+    didSet {
+      DispatchQueue.main.async {
+        self.baseView.tableView.reloadData()
```
좋습니다

jungseungyeo reviewed (approved):
태엔젤, 두두 고생하셨습니다. 
추가적으로 수정해야 되는 부분은 코멘트 남겨두었습니다. 다음 스텝에서 진행해주세요 :) 

확실히 코드베이스의 코드가 보기 좋아졌고 mvc패턴을 잘 지키고 있는거 같아요 ㅎㅎ
